Bricks/strings/piece.h

```diff
@@ -0,0 +1,201 @@
+//         which is noticeably more efficient than a "regular" lexicographical comparison in O(N).
+//         Yes, it moves the world forward, saves trees and stops, or at least delays, global warming.
+//         Thus, non-metaphorically, the description above makes total sense. Keep in mind though, that
+//         making sense did not help C+= survive. At same same time, although contradictory to most predominant
+//         dogmas, I personally stick with the belief that not only it's important for the things to make sense,
+//         but that making sense is the very only property of things that makes them important. -- D.K.
+//
+// Note 3: `Piece` is a better name than `SlaveString`, right?
+
+#ifndef BRICKS_STRINGS_PIECE_H
+#define BRICKS_STRINGS_PIECE_H
+
+#include "../port.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <unordered_map>
+#include <string>
+
+namespace bricks {
+namespace strings {
+
+struct Piece {
+  const char* S;
+  size_t N;
+
+  Piece() : S(""), N(0u) {}
+  Piece(const char* s, size_t n) : S(s), N(n) { assert(S[N] == '\0'); }
+  Piece(const char* s) : S(s), N(strlen(s)) {}
+  template <int L>
+  Piece(const char s[L])
+      : S(s), N(L - 1) {
+    // The above line break is `clang-format`, not me. -- D.K.
+    assert(S[N] == '\0');
+  }
+  Piece(const std::string& s) : S(s.data()), N(s.size()) {}
+  // Copyable and assignable by design.
+
+  bool empty() const { return N == 0u; }
+  size_t length() const { return N; }
+
+  const char* c_str() const { return S; }
+
+  void clear() {
+    S = "";
+    N = 0u;
+  }
+
+  void CheckPrivilege(size_t i) const { assert(i <= N); }
+
+  char operator[](size_t i) const {
+    CheckPrivilege(i);  // Once is enough.
+    return S[i];
+  }
+
+  operator std::string() const { return std::string(S, N); }
+
+  bool HasPrefix(const Piece& rhs) const {
+    const Piece& lhs = *this;
+    return lhs.N >= rhs.N && !::memcmp(lhs.S, rhs.S, rhs.N);
+  }
+
+  int LexicographicalCompare(const Piece& rhs) const {
+    const Piece& lhs = *this;
+    const int d = ::memcmp(S, rhs.S, std::min(lhs.N, rhs.N));
+    if (d) {
+      return d;
+    } else {
+      if (lhs.N < rhs.N) {
+        return -1;
+      } else if (lhs.N > rhs.N) {
+        return +1;
+      } else {
+        return 0;
+      }
+    }
+  }
+
+  struct HashFunction final {
+    // Must actually consider the eternal nature of `Piece`, not just its pointer. Sigh. Inequality in its best.
+    size_t operator()(const Piece& piece) const {
+      // TODO(dkorolev): Use a better hash one day.
+      double hash = 0.0;
+      double k = 1.0;
+      for (size_t i = 0; i < piece.N; ++i, k = cos(k + i)) {
+        hash += k * piece.S[i];
+      }
+      static_assert(sizeof(double) >= sizeof(size_t), "Suddenly, `reinterpet_cast<>` doesn't nail it.");
+      return *reinterpret_cast<size_t*>(&hash);
```

Comment:
While ANSI regulates how exactly `double` laid out, it's but, but it's not important -- the important thing is that this hash function returns the same `size_t` values for same strings, and, hopefully very likely, different values of `size_t` for different strings.